libevmasm/AssemblyItem.cpp

```diff
@@ -184,6 +192,10 @@ size_t AssemblyItem::arguments() const
 		return std::get<0>(*m_verbatimBytecode);
 	else if (type() == AssignImmutable)
 		return 2;
+	else if (type() == EOFCreate)
+		return 4;
+	else if (type() == ReturnContract)
+		return 2;
```

Comment:
Looking at #15534 I realized that here we should have actually used the instruction info, since for the new assembly item types we actually do have it:

```C++
	if (
		type() == Operation ||
		type() == AudDataLoadN ||
		type() == EOFCreate ||
		type() == ReturnContract
	)
		// The latest EVMVersion is used here, since the InstructionInfo is assumed to be
		// the same across all EVM versions except for the instruction name.
		return static_cast<size_t>(instructionInfo(instruction(), EVMVersion()).args);
```

Also in `nameAndData()` and `operator<<()`, at least for `EOFCREATE` and `RETURNCONTRACT`, because they have their own item types only due to the immediate argument and otherwise would just be an `Operation`.

Reply:
First fixed but regarding the second it requires adding you `AssemblyItem` which accepts `AssemblyItemType` and `Instruction`.

Reply:
It would help a lot if we had an easy distinction  between `AssemblyItemType`which has `m_instruction` properly set or not. Having this member properly set we still have to manually filter by item type like here i.e.:
```cpp
bool SemanticInformation::altersControlFlow(AssemblyItem const& _item)
{
	if (_item.type() != evmasm::Operation ||
	    _item.type() != evmasm::ReturnContract)
		return false;

	switch (_item.instruction())
	{
	// note that CALL, CALLCODE and CREATE do not really alter the control flow, because we
	// continue on the next instruction
	case Instruction::JUMP:
	case Instruction::JUMPI:
	case Instruction::RETURN:
	case Instruction::SELFDESTRUCT:
	case Instruction::STOP:
	case Instruction::INVALID:
	case Instruction::REVERT:
	case Instruction::RETURNCONTRACT:
		return true;
	default:
		return false;
	}
}
```
So my idea is to add a static helper `hasInstruction` to `AssemblyItem`

Reply:
I added new constructor and adjusted `instruction()` function accordingly.

Reply:
Yes, a unified check like that sounds like a good idea. But not sure if `isSingleInstruction()`. Some assembly items currently are already single instructions (e.g. `PUSH` and `DUP`. How about just `hasInstruction()`?